src/Features/Core/Portable/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,337 @@
+                    symbolStartContext.RegisterOperationAction(compilationAnalyzer.AnalyzeFieldInitializer, OperationKind.FieldInitializer);
+                    symbolStartContext.RegisterOperationAction(compilationAnalyzer.AnalyzeInvocationOperation, OperationKind.Invocation);
+                    symbolStartContext.RegisterOperationAction(_ => hasInvalidOperation = true, OperationKind.Invalid);
+                    symbolStartContext.RegisterSymbolEndAction(symbolEndContext => compilationAnalyzer.OnSymbolEnd(symbolEndContext, hasInvalidOperation));
+                }, SymbolKind.NamedType);
+            });
+        }
+
+        private sealed class CompilationAnalyzer
+        {
+            private readonly DiagnosticDescriptor _removeUnusedMembersRule, _removeUnreadMembersRule;
+            private readonly object _gate;
+            private readonly Dictionary<ISymbol, ValueUsageInfo> _symbolValueUsageStateMap;
+            private readonly Lazy<INamedTypeSymbol> _lazyTaskType, _lazyGenericTaskType;
+
+            public CompilationAnalyzer(Compilation compilation, DiagnosticDescriptor removeUnusedMembersRule, DiagnosticDescriptor removeUnreadMembersRule)
+            {
+                _removeUnusedMembersRule = removeUnusedMembersRule;
+                _removeUnreadMembersRule = removeUnreadMembersRule;
+                _gate = new object();
+
+                // State map for candidate member symbols, with the value indicating how each symbol is used in executable code.
+                _symbolValueUsageStateMap = new Dictionary<ISymbol, ValueUsageInfo>();
+
+                _lazyTaskType = new Lazy<INamedTypeSymbol>(() => compilation.TaskType());
+                _lazyGenericTaskType = new Lazy<INamedTypeSymbol>(() => compilation.TaskOfTType());
+            }
+
+            public void AnalyzeSymbolDeclaration(SymbolAnalysisContext symbolContext)
+            {
+                if (IsCandidateSymbol(symbolContext.Symbol))
+                {
+                    lock (_gate)
+                    {
+                        // Initialize unused state to 'true'.
+                        // Note that we might receive a symbol reference (AnalyzeMemberOperation) callback before
+                        // this symbol declaration callback, so even though we cannot receive duplicate callbacks for a symbol,
+                        // an entry might already be present of the declared symbol here.
+                        if (!_symbolValueUsageStateMap.ContainsKey(symbolContext.Symbol))
+                        {
+                            _symbolValueUsageStateMap.Add(symbolContext.Symbol, ValueUsageInfo.None);
+                        }
+                    }
+                }
+            }
+
+            public void AnalyzeFieldInitializer(OperationAnalysisContext operationContext)
```

Comment:
as per above, these coudl all be private if CompilationAnalyzer is responsible for registering callbacks.